File: AdPieSample/AdPieSample/Source/UIViewController+Toast.m
#import "UIViewController+Toast.h"

@implementation UIViewController (Toast)

- (UIViewController * _Nullable)topMostViewController {
    UIWindowScene *activeScene = nil;
    
    if (@available(iOS 13.0, *)) {
        for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
            if (scene.activationState == UISceneActivationStateForegroundActive && [scene isKindOfClass:[UIWindowScene class]]) {
                activeScene = (UIWindowScene *)scene;
                break;
            }
        }
    }
    
    UIWindow *keyWindow = nil;
    if (@available(iOS 13.0, *)) {
        for (UIWindow *window in activeScene.windows) {
            if (window.isKeyWindow) {
                keyWindow = window;
                break;
            }
        }
    } else {
        keyWindow = UIApplication.sharedApplication.keyWindow;
    }
    
    if (!keyWindow) {
        return nil;
    }
    
    UIViewController *topController = keyWindow.rootViewController;
    while (topController.presentedViewController) {
        topController = topController.presentedViewController;
    }
    
    return topController;
}

- (void)showToastWithMessage:(NSString *)message {
    UIView *containerView = [[UIView alloc] init];
    containerView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.7];
    containerView.layer.cornerRadius = 16.0;
    containerView.clipsToBounds = YES;
    containerView.translatesAutoresizingMaskIntoConstraints = NO;
    containerView.alpha = 0.0;
    
    UILabel *label = [[UILabel alloc] init];
    label.textColor = [UIColor whiteColor];
    label.textAlignment = NSTextAlignmentCenter;
    label.font = [UIFont systemFontOfSize:14.0 weight:UIFontWeightMedium];
    label.text = message;
    label.numberOfLines = 0;
    label.translatesAutoresizingMaskIntoConstraints = NO;
    
    [containerView addSubview:label];
    
    UIViewController *topVC = [self topMostViewController];
    UIView *targetView = topVC ? topVC.view : self.view;
    
    if (!targetView) {
        return;
    }
    
    [targetView addSubview:containerView];
    
    [NSLayoutConstraint activateConstraints:@[
        [containerView.centerXAnchor constraintEqualToAnchor:targetView.centerXAnchor],
        [containerView.bottomAnchor constraintEqualToAnchor:targetView.safeAreaLayoutGuide.bottomAnchor constant:-100],
        [containerView.leadingAnchor constraintGreaterThanOrEqualToAnchor:targetView.leadingAnchor constant:20],
        [containerView.trailingAnchor constraintLessThanOrEqualToAnchor:targetView.trailingAnchor constant:-20],
        
        [label.topAnchor constraintEqualToAnchor:containerView.topAnchor constant:10],
        [label.bottomAnchor constraintEqualToAnchor:containerView.bottomAnchor constant:-10],
        [label.leadingAnchor constraintEqualToAnchor:containerView.leadingAnchor constant:16],
        [label.trailingAnchor constraintEqualToAnchor:containerView.trailingAnchor constant:-16]
    ]];
    
    [UIView animateWithDuration:0.3 animations:^{
        containerView.alpha = 1.0;
    } completion:^(BOOL finished) {
        [UIView animateWithDuration:0.3 delay:2.0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            containerView.alpha = 0.0;
        } completion:^(BOOL finished) {
            [containerView removeFromSuperview];
        }];
    }];
}

@end

File: AdPieSwiftSample/AdPieSwiftSample/Source/UIViewController+Toast.swift
import UIKit

extension UIViewController {
    private var topMostViewController: UIViewController? {
        guard let windowScene = UIApplication.shared.connectedScenes.first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene,
              let keyWindow = windowScene.windows.first(where: { $0.isKeyWindow }) else {
            return nil
        }
        var topController = keyWindow.rootViewController
        while let presentedViewController = topController?.presentedViewController {
            topController = presentedViewController
        }
        return topController
    }

    func showToast(message: String) {
        let containerView = UIView()
        containerView.backgroundColor = UIColor.black.withAlphaComponent(0.7)
        containerView.layer.cornerRadius = 16
        containerView.clipsToBounds = true
        containerView.translatesAutoresizingMaskIntoConstraints = false
        containerView.alpha = 0.0
        
        let label = UILabel()
        label.textColor = .white
        label.textAlignment = .center
        label.font = .systemFont(ofSize: 14.0, weight: .medium)
        label.text = message
        label.numberOfLines = 0
        label.translatesAutoresizingMaskIntoConstraints = false
        
        containerView.addSubview(label)
        
        guard let targetView = topMostViewController?.view ?? self.view else { return }
        targetView.addSubview(containerView)
        
        NSLayoutConstraint.activate([
            containerView.centerXAnchor.constraint(equalTo: targetView.centerXAnchor),
            containerView.bottomAnchor.constraint(equalTo: targetView.safeAreaLayoutGuide.bottomAnchor, constant: -100),
            containerView.leadingAnchor.constraint(greaterThanOrEqualTo: targetView.leadingAnchor, constant: 20),
            containerView.trailingAnchor.constraint(lessThanOrEqualTo: targetView.trailingAnchor, constant: -20),
            
            label.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 10),
            label.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: -10),
            label.leadingAnchor.constraint(equalTo: containerView.leadingAnchor, constant: 16),
            label.trailingAnchor.constraint(equalTo: containerView.trailingAnchor, constant: -16)
        ])
        
        UIView.animate(withDuration: 0.3, animations: {
            containerView.alpha = 1.0
        }) { _ in
            UIView.animate(withDuration: 0.3, delay: 2.0, options: .curveEaseOut, animations: {
                containerView.alpha = 0.0
            }, completion: { _ in
                containerView.removeFromSuperview()
            })
        }
    }
}
